### ISSUE TITLE
Update to database hangs without exception

### ISSUE THREAD

**Issue by AnvarYasinov1997** [jOOQ/jOOQ#14491]
### Expected behavior

I am using org.jooq:jooq:3.17.4 in conjunction with org.postgresql:r2dbc-postgresql:0.9.2.RELEASE
As well as
```
implementation("io.r2dbc:r2dbc-spi:1.0.0.RELEASE")
implementation("io.r2dbc:r2dbc-pool:0.9.2.RELEASE")
```

To implement connection pooling

When updating data in the database, my thread freezes with no signs of life in the logs

      // on calling this method the code freze
    suspend fun doneByIds(ids: List<UUID>) {
        monoDslContext { dslContext ->
            dslContext.update(OUTBOX)
                .set(OUTBOX.STATUS, DONE)
                .where(OUTBOX.ID.`in`(ids))
                .and(OUTBOX.STATUS.eq(PROCESSING))
                .toMono()
        }.awaitFirst()
    }

      suspend fun setProcessingByIds(ids: List<UUID>): List<Outbox> =
        monoDslContext { dslContext ->
            dslContext.update(OUTBOX)
                .set(OUTBOX.STATUS, PROCESSING)
                .where(OUTBOX.ID.`in`(ids))
                .and(OUTBOX.STATUS.eq(AWAIT))
                .returning()
                .toFlux()
                .map(::Outbox)
                .collectList()
        }.awaitFirst()

    suspend fun findAllIdsAwaitedByType(type: Type, limit: Long): List<UUID> =
        monoDslContext { dslContext ->
            dslContext.select(OUTBOX.ID)
                .from(OUTBOX)
                .where(OUTBOX.TYPE.eq(type.name))
                .and(OUTBOX.STATUS.eq(AWAIT))
                .limit(limit)
                .toFlux()
                .map(Record1<UUID>::value1)
                .collectList()
        }.awaitFirst()

     //configure scheduled task with use springboot SchedulingConfigurer class
    override fun configureTasks(taskRegistrar: ScheduledTaskRegistrar) {
        taskRegistrar.addFixedDelayTask(getTask(), getTriggeredFrequencyMillis())
    }

      fun getTask(): Runnable = Runnable {
        runBlocking {
            val idsToProcess = outboxRepository.findAllIdsAwaitedByType(type, getLimit())
            val outboxList = outboxRepository.setProcessingByIds(idsToProcess)
            val ids = outboxList.map(Outbox::getId)

            log("Notifications sent")
            //Here is the method call
            outboxRepository.doneByIds(ids)
            log("Notifications are set to DONE status")
        }
       }

log("Notifications sent") 
this is the last log from the thread that I see in the console, while I wrapped the code in a try block, and no exceptions are thrown

Here is my dslContext getting method

    import org.jooq.Configuration

    private val successBlock: (connection: Connection) -> Mono<*> =
        { connection -> commitTransactionAndCloseConnection(connection) }
    private val errorBlock: (connection: Connection, e: Throwable) -> Mono<*> =
        { connection, _ -> rollbackTransactionAndCloseConnection(connection) }
    private val cancelBlock: (connection: Connection) -> Mono<*> =
        { connection -> rollbackTransactionAndCloseConnection(connection) }

    protected fun <T> monoDslContext(config: Configuration? = null, block: (dslContext: DSLContext) -> Mono<T>): Mono<T> {
        if (config != null) return block(config.dsl())
        return useMonoDslContext(block)
    }

    fun <T> useMonoDslContext(block: (dslContext: DSLContext) -> Mono<T>): Mono<T> {
        val dslContextBlock: (connection: Connection) -> Mono<T> = { connection -> block(DSL.using(connection)) }
        return Mono.usingWhen(connectionPool.create(), dslContextBlock, successBlock, errorBlock, cancelBlock)
    }

    private fun rollbackTransactionAndCloseConnection(connection: Connection): Mono<*> =
        Mono.from(connection.rollbackTransaction())
            .then(Mono.from(connection.close()))

    private fun commitTransactionAndCloseConnection(connection: Connection): Mono<*> =
        Mono.from(connection.commitTransaction())
            .then(Mono.from(connection.close()))`

As well as pool initialization methods

     @Bean
     fun connectionFactory(): ConnectionFactory =
        PostgresqlConnectionConfiguration.builder()
            .host(properties.host)
            .port(properties.port)
            .username(properties.username)
            .password(properties.password)
            .database(properties.dbName)
            .build()
            .let(::PostgresqlConnectionFactory)

    @Bean
    fun connectionPool(connectionFactory: ConnectionFactory): ConnectionPool =
        ConnectionPoolConfiguration.builder(connectionFactory)
            .maxIdleTime(Duration.ofMinutes(2))
            .maxLifeTime(Duration.ofMinutes(1))
            .backgroundEvictionInterval(Duration.ofSeconds(5))
            .maxCreateConnectionTime(Duration.ofSeconds(3))
            .maxValidationTime(Duration.ofSeconds(3))
            .maxAcquireTime(Duration.ofSeconds(3))
            .initialSize(properties.initialPoolSize) //10
            .maxSize(properties.maxPoolSize) //50
            .build()
            .let(::ConnectionPool)`

I have four schedulers that work in parallel and parse the queue, the trigger is once every 1 second. When one of the 4 threads freezes, the remaining 3 continue to work, while they work with the same "outbox" table without using transactions

### Steps to reproduce the problem

The problem reproduces in random order, that is, my service can work for a week without errors and break after 7 days after launch, the same problem can occur an hour after launch

### jOOQ Version

3.17.4

### Database product and version

14.5-alpine

### Java Version

17

### OS Version

Ubuntu 20.04.3 LTS

### JDBC driver name and version (include name if unofficial driver)

org.postgresql:r2dbc-postgresql:0.9.2.RELEASE

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'm currently reviewing a few R2DBC related bugs, I might be able to look into this. Though, if there's any way to reliably reproduce this, that would be helpful. Flaky problems are hard to spot. We currently don't have any hanging processes in our integration tests (though there have been, in the past).

One issue that immediately comes to mind is that you're using the 1.0 version of the SPI and a 0.9 version of the driver. There are incompatibilities, and jOOQ 3.17 doesn't yet support the 1.0 SPI version, see https://github.com/jOOQ/jOOQ/issues/13565. It will be supported only starting with jOOQ 3.18

**Comment by AnvarYasinov1997** [comment]
Unfortunately, I can't give you a sample with a consistently reproducible bug, since the bug is only reproducible in production environments and with unknown frequency. We wrote logs before and after each call in a non-working block, the hang occurs randomly, regardless of the number of records being updated, and regardless of the thread processing the record type in the database.
By the way, this problem does not get out on the staging environment, I think because there is no large load on the database, limited by a pool of 50 connections.

**Comment by lukaseder** [comment]
I know it's hard to reproduce such asynchronous bugs, but without clear evidence what's causing the problem, it will be hard to write a regression test and be certain of a fix.

In the past, the problem appeared when some exception was thrown at the "wrong spot", and instead of catching it and signalling `onError`, it leaked into reactor, which then didn't cancel subscriptions, etc. Does the problem persist with the latest jOOQ version? Do you see any indication of such problems in the logs?

**Comment by lukaseder** [comment]
Note, I've just fixed https://github.com/jOOQ/jOOQ/issues/14489 in jOOQ 3.18.0 and 3.17.7, which may or may not be related. In that issue, an `R2dbcException` could lead to connections not being closed. The problem manifested only in SQL Server in our integration tests, but it may well be that it appears as well elsewhere, under load.

**Comment by AnvarYasinov1997** [comment]
the problem is that I have no exceptions, I subscribe to mono through awaitFirst, the program falls asleep forever, and the scheduler thread does not show signs of life

**Comment by lukaseder** [comment]
Yeah, those problems seem to be typical when blocking / suspending on a signal that is supposed to be sent, but then it never is, because of a bug.

Another issue that got fixed last week is https://github.com/jOOQ/jOOQ/issues/13802. It is related to wrong handling of `Mono` related signals in transactions, which is what you're doing. In particular, this comment explains the problem: https://github.com/jOOQ/jOOQ/issues/13802#issuecomment-1381990333

In short, reactor sends a `Subscription::cancel` signal *before* the `Subscriber::onNext` signal, meaning the connection may be closed before the expected record / whatever is sent to the client. But that client is still waiting for the record, forever.

I'll release 3.17.7 this week. Without a minimal reproducer, maybe you could try upgrading to this version and check if the problem persists? You could build `3.17.7-SNAPSHOT` and try it already now.

**Comment by AnvarYasinov1997** [comment]
I received your latest release patch, thanks for your support)))

**Comment by lukaseder** [comment]
> I received your latest release patch, thanks for your support)))

With that latest patch version, can you still reproduce this problem?

**Comment by lukaseder** [comment]
I'm assuming this has been fixed now with one of the recent releases, which included fixes for this kind of R2DBC problem. Please let me know if it still persists.
